demos/jobs/jobs_demo_mosquitto/jobs_demo_mosquitto.c

```diff
@@ -0,0 +1,1304 @@
+                                '.',
+                                &url,
+                                &urlLength );
+    }
+
+    if( json_ret == JSONSuccess )
+    {
+        jobid = strndup( jobid, jobidLength );
+        url = strndup( url, urlLength );
+        assert( ( jobid != NULL ) && ( url != NULL ) );
+
+        h->jobid = jobid;
+        h->jobidLength = jobidLength;
+        h->url = url;
+        h->urlLength = urlLength;
+        ret = true;
+    }
+    else
+    {
+        if( jobid != NULL )
+        {
+            jobid[ jobidLength ] = '\0';
+            warnx( "missing url; failing job id: %s", jobid );
+            ( void ) sendUpdate( h, jobid, jobidLength, makeReport_( "FAILED" ) );
+        }
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+void on_message( struct mosquitto * m,
+                 void * p,
+                 const struct mosquitto_message * message )
+{
+    handle_t * h = p;
+    JobsStatus_t ret;
+    JobsTopic_t api;
+    char * jobid;
+    uint16_t jobidLength;
+
+    assert( h != NULL );
+    assert( message->topic != NULL );
+
+    /* identify a Jobs API and a job ID represented in a topic buffer */
+    ret = Jobs_MatchTopic( message->topic,
+                           strlen( message->topic ),
+                           h->name,
+                           h->nameLength,
+                           &api,
+                           &jobid,
+                           &jobidLength );
+    assert( ret != JobsBadParameter );
+    ( void ) ret;
+
+    switch( api )
+    {
+        /* a job has been added or the current job was canceled */
+        case JobsNextJobChanged:
+            h->lastPrompt = 0;
+            h->lastUpdate = 0;
+            break;
+
+        /* response to a request to start the next job */
+        case JobsStartNextSuccess:
+
+            if( h->runStatus == None )
+            {
+                if( parseJob( h, message ) == true )
+                {
+                    h->runStatus = Ready;
+                }
+            }
+            else
+            {
+                warnx( "unexpected start next message" );
+            }
+
+            break;
+
+        /* The last update was rejected. */
+        case JobsUpdateFailed:
+
+            if( ( h->runStatus == Running ) &&
+                ( h->jobid != NULL ) &&
+                ( h->jobidLength == jobidLength ) &&
+                ( strncmp( h->jobid, jobid, jobidLength ) == 0 ) )
+            {
+                h->runStatus = Cancel;
+            }
+            else
+            {
+                jobid[ jobidLength ] = '\0';
+                warnx( "update failure for unknown job id: %s", jobid );
+
+                if( ( message->payload != NULL ) && ( message->payloadlen > 0 ) )
+                {
+                    char * buffer = message->payload;
+                    buffer[ message->payloadlen ] = '\0';
```

Comment:
Is it a possible out-of-bounds write?

Reply:
Probably.  I removed this, as it was a leftover debugging aid.